### PR TITLE
feat: backend support for inference metric tracking part 1

### PR DESCRIPTION
**Pull request by tayritenour**
## Description

[MLG-637]

Design Doc: https://hpe.sharepoint.com/:w:/t/detai/EVAlau0Y2exMgXyX63inrmEB7rUFHEfB8bDUmdLTxJ4Bjw?e=ZjAZJ7&clickparams=eyJBcHBOYW1lIjoiVGVhbXMtRGVza3RvcCIsIkFwcFZlcnNpb24iOiIyOC8yMzA1MDEwMDQyMiIsIkhhc0ZlZGVyYXRlZFVzZXIiOmZhbHNlfQ%3D%3D

Adds the ability to keep track of what checkpoints were used in a given trial. When leveraged with the new generic metrics APIs, this will allow us to track and find the metrics for all inference runs that utilize a given checkpoint.

Related PRs: 

- https://github.com/determined-ai/determined/pull/7117
- https://github.com/determined-ai/determined/pull/7274

## Test Plan

The integration tests pass. This isn't integrated fully for the user to see in a release party.

## Checklist

- [ ] Changes have been manually QA'd
- [ ] User-facing API changes need the "User-facing API Change" label.
- [ ] Release notes should be added as a separate file under `docs/release-notes/`.
  See [Release Note](https://github.com/determined-ai/determined/blob/master/docs/release-notes/README.md) for details.
- [ ] Licenses should be included for new code which was copied and/or modified from any external code.

## Ticket

[MLG-637]: https://hpe-aiatscale.atlassian.net/browse/MLG-637?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: proto/src/determined/trial/v1/trial.proto [hunk 1]
```diff
@@ -276,3 +276,33 @@ message MetricsReport {
...
+    }
+  };
+  // ID of the trial.
+  int32 trial_id = 1;
+  // UUID of the checkpoint.
+  string checkpoint_uuid = 2;
+  // Source model_version id which generated the checkpoint (if applicable)
+  int32 source_model_version_id = 3;
+  // Source model_version version which generated the checkpoint (if applicable)
+  int32 source_model_version_version = 4;
```
Comment (ioga) [inline proto/src/determined/trial/v1/trial.proto:305]: seems like these things are optional. can we:

```suggestion
  // Source model_version id which generated the checkpoint (if applicable)
  optional int32 source_model_version_id = 3;
  // Source model_version version which generated the checkpoint (if applicable)
  optional int32 source_model_version_version = 4;
```

2. Whenever we check if they are present, instead of checking they're 0, check for nil, i.e.

```
if tsi.SourceModelVersionId != nil {
```

File: master/internal/api_trials.go [hunk 2]
```diff
@@ -912,6 +912,68 @@ func (a *apiServer) streamMetrics(ctx context.Context,
...
+	tsi := req.TrialSourceInfo
+	query := db.Bun().NewInsert().Model(tsi).
+		Value("trial_source_info_type", "?", tsi.TrialSourceInfoType.String()).
+		Returning("trial_id").Returning("checkpoint_uuid")
+	if tsi.GetSourceModelVersionId() == 0 {
+		query.ExcludeColumn("source_model_version_id")
+	}
+	if tsi.GetSourceModelVersionVersion() == 0 {
+		query.ExcludeColumn("source_model_version_version")
+	}
```
Comment (ioga) [inline master/internal/api_trials.go:931]: commented on proto: these fields should be made optional, and then 
```suggestion
	if tsi.SourceModelVersionId == nil {
		query.ExcludeColumn("source_model_version_id")
	}
	if tsi.SourceModelVersionVersion == nil {
		query.ExcludeColumn("source_model_version_version")
	}
```

File: master/internal/api_trials.go [hunk 3]
```diff
@@ -912,6 +912,68 @@ func (a *apiServer) streamMetrics(ctx context.Context,
 	return nil
 }
 
+// Create a TrialSourceInfo, which serves as a link between trials and checkpoints
+// used for tracking purposes for fine tuning and inference.
+func (a *apiServer) ReportTrialSourceInfo(
+	ctx context.Context, req *apiv1.ReportTrialSourceInfoRequest,
+) (*apiv1.ReportTrialSourceInfoResponse, error) {
+	// TODO (Taylor): Handle user auth/rbac
```
Comment (ioga) [inline master/internal/api_trials.go:920]: yep, probably `CanEditExperiment`.

File: master/internal/api_trials.go [hunk 4]
```diff
@@ -912,6 +912,68 @@ func (a *apiServer) streamMetrics(ctx context.Context,
 	return nil
 }
 
+// Create a TrialSourceInfo, which serves as a link between trials and checkpoints
```
Comment (ioga) [inline master/internal/api_trials.go:915]: since these aren't using any of `apiServer` properties, it'd be great to move them to `internal/trials/api_trials.go` which we used to have before [the trial collections removal](https://github.com/determined-ai/determined/pull/7297/files#diff-906e5e11d91f28bf7359060528b003d91b6b50f28491ebbbdafe83dc0a7bc388) and [will have again](https://github.com/determined-ai/determined/pull/7060/files#diff-906e5e11d91f28bf7359060528b003d91b6b50f28491ebbbdafe83dc0a7bc388) in detached mode?

basically: 
- add `type TrialsAPIServer struct{}`
- embed it into the apiserver in `api.go`
- move these methods onto `TrialsAPIServer`.

File: master/internal/api_trials.go [hunk 5]
```diff
@@ -912,6 +912,68 @@ func (a *apiServer) streamMetrics(ctx context.Context,
...
+
+// Query for all trials that use a given checkpoint and return their metrics.
+func (a *apiServer) GetTrialSourceInfoMetricsByCheckpoint(
+	ctx context.Context, req *apiv1.GetTrialSourceInfoMetricsByCheckpointRequest,
+) (*apiv1.GetTrialSourceInfoMetricsByCheckpointResponse, error) {
+	// TODO (Taylor): Handle user auth/rbac
+	resp := &apiv1.GetTrialSourceInfoMetricsByCheckpointResponse{}
+	trialIDsQuery := db.Bun().NewSelect().Table("trial_source_infos").
+		Where("checkpoint_uuid = ?", req.CheckpointUuid).
+		Column("trial_id").Column("trial_source_info_type")
```
Comment (ioga) [inline master/internal/api_trials.go:944]: you can also do this, which tends to be more readable when you have a ton of columns: 

```suggestion
		Column("trial_id", "trial_source_info_type")
```

File: master/internal/api_trials.go [hunk 6]
```diff
@@ -912,6 +912,68 @@ func (a *apiServer) streamMetrics(ctx context.Context,
...
+	}
+	if tsi.GetSourceModelVersionVersion() == 0 {
+		query.ExcludeColumn("source_model_version_version")
+	}
+	_, err := query.Exec(ctx, resp)
+	return resp, err
+}
+
+// Query for all trials that use a given checkpoint and return their metrics.
+func (a *apiServer) GetTrialSourceInfoMetricsByCheckpoint(
```
Comment (ioga) [inline master/internal/api_trials.go:937]: naming nitpick: perhaps call it `GetTrialMetricsBySourceInfoCheckpoint`?.. I feel like that'd better reflect the relationship?
`TrialSourceInfoMetrics` object also feels strange, because it's not metrics on `trial_source_into`, it's the metrics from a trial obtained through a relationship in `trial_source_info`.

File: proto/src/determined/trial/v1/trial.proto (hunk 1 above)
Comment (tayritenour) [reply proto/src/determined/trial/v1/trial.proto:305]: When I compile with `optional` I get this result:

```go
func (x *TrialSourceInfo) GetSourceModelVersionId() int32 {
	if x != nil && x.SourceModelVersionId != nil {
		return *x.SourceModelVersionId
	}
	return 0
}
```

Comment (ioga) [reply proto/src/determined/trial/v1/trial.proto:305]: use the property `tsi.SourceModelVersionId` itself, not the getter method.

File: harness/determined/determined.code-workspace [hunk 7]
```diff
@@ -0,0 +1,7 @@
+{
```
Comment (ioga) [inline harness/determined/determined.code-workspace:1]: prob remove this file

File: master/internal/trials/api_trial_source_info.go [hunk 8]
```diff
@@ -0,0 +1,82 @@
...
+	q = q.Column("trial_id", "trial_source_info_type")
+	err := q.Scan(ctx, &trialIds)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get trial source info %w", err)
+	}
+
+	ret := []*apiv1.TrialSourceInfoMetric{}
+	numMetricsLimit := 1000
+	for _, val := range trialIds {
+		if err := CanGetTrialsExperimentAndCheckCanDoAction(ctx, val.TrialID,
```
Comment (ioga) [inline master/internal/trials/api_trial_source_info.go:62]: if there's a ton of trial ids, it may be more optimal to join trial_source_info to trials to experiments, get workspace ids, and check permissions once per workspace id.
I won't insist on adding it right now, but perhaps as a TODO or something to keep an eye on from perf perspective.

File: proto/src/determined/api/v1/trial.proto [hunk 9]
```diff
@@ -797,7 +797,29 @@ message GetTrialSourceInfoMetricsByCheckpointRequest {
...
+  repeated TrialSourceInfoMetric data = 1;
+}
+
+// Request for all metrics related to a given checkpoint
+message GetTrialSourceInfoMetricsByModelVersionRequest {
+  option (grpc.gateway.protoc_gen_swagger.options.openapiv2_schema) = {
+    json_schema: { required: [ "checkpoint_uuid" ] }
+  };
+  // The database id of the model version
+  int32 model_version_id = 1;
```
Comment (ioga) [inline proto/src/determined/api/v1/trial.proto:814]: hmmm, isn't this a model_id? not model_version_id.

File: proto/src/determined/trial/v1/trial.proto [hunk 10]
```diff
@@ -300,9 +300,9 @@ message TrialSourceInfo {
   // UUID of the checkpoint.
   string checkpoint_uuid = 2;
   // Source model_version id which generated the checkpoint (if applicable)
-  int32 source_model_version_id = 3;
+  optional int32 model_version_id = 3;
```
Comment (ioga) [inline proto/src/determined/trial/v1/trial.proto:303]: same: isn't this a model_id?

File: proto/src/determined/api/v1/trial.proto (hunk 9 above)
Comment (tayritenour) [reply proto/src/determined/api/v1/trial.proto]: Changed this to better reflect where the id's are actually coming from. It's now `model_id` and `model_version` where they point to the `Model.Id` and `ModelVersion.Version` respectively

File: master/internal/trials/api_trial_source_info.go (hunk 8 above)
Comment (tayritenour) [reply master/internal/trials/api_trial_source_info.go:65]: Added a comment about this

**Review by ioga — APPROVED** [review]

File: master/internal/trials/api_trial_source_info.go (hunk 8 above)
Comment (hamidzr) [reply master/internal/trials/api_trial_source_info.go:65]: do we also usually want a ticket associated with these inline todos?

File: master/internal/trials/api_trial_source_info.go [hunk 11]
```diff
@@ -0,0 +1,85 @@
...
+
+// ReportTrialSourceInfo creates a TrialSourceInfo, which serves as a link between
+// trials and checkpoints used for tracking purposes for fine tuning and inference.
+func (a *TrialSourceInfoAPIServer) ReportTrialSourceInfo(
+	ctx context.Context, req *apiv1.ReportTrialSourceInfoRequest,
+) (*apiv1.ReportTrialSourceInfoResponse, error) {
+	tsi := req.TrialSourceInfo
+	if err := CanGetTrialsExperimentAndCheckCanDoAction(ctx, int(tsi.TrialId),
+		experiment.AuthZProvider.Get().CanEditExperiment); err != nil {
+		return nil, err
```
Comment (hamidzr) [inline master/internal/trials/api_trial_source_info.go:28]: we want to double-check that this doesn't give away that a trial exists to a caller who shouldn't have view access. Have you tested that?

Comment (hamidzr) [reply master/internal/trials/api_trial_source_info.go:28]: applies anywhere else we're consulting our RBAC

File: master/internal/api_model.go [hunk 12]
```diff
@@ -813,3 +814,39 @@ func (a *apiServer) DeleteModelVersion(
...
+	if err != nil {
+		return nil, err
+	}
+	curUser, _, err := grpcutil.GetUser(ctx)
+	if err != nil {
+		return nil, err
+	}
+	if err := modelauth.AuthZProvider.Get().CanGetModel(ctx, *curUser, modelResp.Model,
+		modelResp.Model.WorkspaceId); err != nil {
+		return nil, err
```
Comment (hamidzr) [inline master/internal/api_model.go:834]: ditto

**Review by hamidzr — COMMENTED** [review]
@tayritenour asked I take a quick look. I commented on some bits that jumped out w/o fully understanding everything.

File: master/internal/trials/utils.go [hunk 13]
```diff
@@ -0,0 +1,44 @@
...
+	"github.com/determined-ai/determined/master/internal/authz"
+	"github.com/determined-ai/determined/master/internal/db"
+	"github.com/determined-ai/determined/master/internal/experiment"
+	"github.com/determined-ai/determined/master/internal/grpcutil"
+	"github.com/determined-ai/determined/master/pkg/model"
+)
+
+// CanGetTrialsExperimentAndCheckCanDoAction is a utility function for generalizing
+// RBAC support for trials and experiments.
+func CanGetTrialsExperimentAndCheckCanDoAction(ctx context.Context,
```
[the remark this answers is not in the document]
Comment (tayritenour) [reply master/internal/trials/utils.go:21]: ah okay I didn't know we had a place for stuff like this. I can look into it. I was just extracting it out of the `api_trials_[...]` file

[the remark this answers is not in the document]
Comment (ioga) [reply master/internal/trials/utils.go:21]: rbac package is the rbac-specific APIs. I don't think we should bundle any permission check helpers for all the things in the system in there.

in other words, I envisioned having `trials` directory contain database code, APIs, permission checks for working with trials; `model` for database/API/permissions for model registry, etc.

File: proto/src/determined/trial/v1/trial.proto [hunk 14]
```diff
@@ -276,3 +276,34 @@ message MetricsReport {
   // Run ID of trial when metric was reported.
   int32 trial_run_id = 7;
 }
+
+// TrialSourceInfoType see \dT+ trial_source_info_type in db
```
[the remark this answers is not in the document]
Comment (tayritenour) [reply proto/src/determined/trial/v1/trial.proto]: I updated this to be more helpful

File: master/internal/trials/api_trial_source_info.go (hunk 11 above)
Comment (tayritenour) [reply master/internal/trials/api_trial_source_info.go:28]: there is now testing for this in `TestTrialSourceInfoCheckpoint`

**Review by hamidzr — APPROVED** [review]
